sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala

```diff
@@ -0,0 +1,155 @@
+
+  val numPartitions = options.get(ContinuousRateStreamSource.NUM_PARTITIONS).orElse("5").toInt
+  val rowsPerSecond = options.get(ContinuousRateStreamSource.ROWS_PER_SECOND).orElse("6").toLong
+
+  override def mergeOffsets(offsets: Array[PartitionOffset]): Offset = {
+    assert(offsets.length == numPartitions)
+    val tuples = offsets.map {
+      case ContinuousRateStreamPartitionOffset(p, s) => p -> s
+    }
+    ContinuousRateStreamOffset(Map(tuples: _*))
+  }
+
+  override def deserializeOffset(json: String): Offset = {
+    ContinuousRateStreamOffset(Serialization.read[Map[Int, Long]](json))
+  }
+
+  override def readSchema(): StructType = {
+    StructType(
+        StructField("timestamp", TimestampType, false) ::
+        StructField("value", LongType, false) :: Nil)
+  }
+
+  private var offset: java.util.Optional[Offset] = _
+
+  override def setOffset(offset: java.util.Optional[Offset]): Unit = {
+    this.offset = offset
+  }
+
+  override def getStartOffset(): Offset = offset.get()
+
+  override def createReadTasks(): java.util.List[ReadTask[Row]] = {
+    val partitionStartMap = Option(offset.orElse(null)).map {
+      case o: ContinuousRateStreamOffset => o.partitionToStartValue
+      case s: SerializedOffset => Serialization.read[Map[Int, Long]](s.json)
+      case _ => throw new IllegalArgumentException("invalid offset type for ContinuousRateSource")
+    }
+    if (partitionStartMap.exists(_.keySet.size > numPartitions)) {
+      throw new IllegalArgumentException("Start offset contained too many partitions.")
+    }
+    val perPartitionRate = rowsPerSecond.toDouble / numPartitions.toDouble
+
+    Range(0, numPartitions).map { n =>
+      // If the offset doesn't have a value for this partition, start from the beginning.
+      val start = partitionStartMap.flatMap(_.get(n)).getOrElse(0L + n)
+      // Have each partition advance by numPartitions each row, with starting points staggered
+      // by their partition index.
+      RateStreamReadTask(start, n, numPartitions, perPartitionRate)
+        .asInstanceOf[ReadTask[Row]]
+    }.asJava
+  }
+
+  override def commit(end: Offset): Unit = {}
+  override def stop(): Unit = {}
+
+}
+
+case class RateStreamReadTask(
+    startValue: Long, partitionIndex: Int, increment: Long, rowsPerSecond: Double)
+  extends ReadTask[Row] {
+  override def createDataReader(): DataReader[Row] =
+    new RateStreamDataReader(startValue, partitionIndex, increment, rowsPerSecond.toLong)
+}
+
+class RateStreamDataReader(
+    startValue: Long, partitionIndex: Int, increment: Long, rowsPerSecond: Long)
+  extends ContinuousDataReader[Row] {
+  private var nextReadTime = 0L
+  private var numReadRows = 0L
+
+  private var currentValue = startValue
+  private var currentRow: Row = null
+
+  override def next(): Boolean = {
+    // Set the timestamp for the first time.
+    if (currentRow == null) nextReadTime = System.currentTimeMillis() + 1000
+
+    if (numReadRows == rowsPerSecond) {
```

Comment:
The timestamp column will be uneven in the current implementation. Could you just calculate how long to sleep for each row and sleep for each row?

Reply:
I could, but the unevenness matches the microbatch behavior. Having it sleep for each row would make it significantly harder to write reliable tests.

Reply:
The microbatch source generates even timestamps. Here is the code: https://github.com/apache/spark/blob/3c0c2d09ca89c6b6247137823169db17847dfae3/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/RateSourceProvider.scala#L213

Reply:
Hm, so it does. I'm not sure why I thought it didn't.

Having said that, I still think even timestamps would pose a serious obstacle to writing reliable tests of the continuous processing engine, which is the primary goal of this implementation. Having a second long window to commit data and shut down the query lets us figure out exactly which rows should have made it to the sink, at least in Spark unit tests where everything is running locally. If the rows come evenly, we'll have a much harder time figuring out what data is supposed to be present.

Reply:
We discussed this offline, and agreed that:

* it's best for the timestamps to be even
* it's fine to adjust the rate source tests in the full implementation to only check that at least the expected data is present (but more may have come in).